File: dotconfig/i3/bin/switch_ws.py
#!/usr/bin/env python3

import json
import subprocess
import re
import sys


def main():
    direction = sys.argv[1].lower()

    def extract_number(ws_name):
        match = re.match(r"^(\d+)", ws_name)
        if match:
            return int(match.group(1))
        return 0

    workspace_data = json.loads(
        subprocess.check_output(["i3-msg", "-t", "get_workspaces"]).decode()
    )

    current_ws = None
    active_output = None

    for workspace in workspace_data:
        if workspace["focused"]:
            current_ws = workspace["name"]
            active_output = workspace["output"]
            break

    workspaces_on_output = []
    for workspace in workspace_data:
        if workspace["output"] == active_output:
            workspaces_on_output.append(workspace["name"])

    workspaces_on_output.sort(key=extract_number)

    current_num = extract_number(current_ws)

    target_ws = None

    if direction == "next":
        for ws in workspaces_on_output:
            if extract_number(ws) > current_num:
                target_ws = ws
                break

        if not target_ws:
            target_ws = workspaces_on_output[0]
    else:
        prev_workspaces = [
            ws for ws in workspaces_on_output if extract_number(ws) < current_num
        ]
        if prev_workspaces:
            target_ws = prev_workspaces[-1]
        else:
            target_ws = workspaces_on_output[-1]

    subprocess.run(["i3-msg", "workspace", target_ws])


if __name__ == "__main__":
    main()

File: dotconfig/i3/bin/switch_ws.pl
#!/usr/bin/env perl

use strict;
use warnings;
use JSON;
use IPC::Open3;
use IO::Handle;

sub extract_number {
    my ($ws_name) = @_;
    if ($ws_name =~ /^(\d+)/) {
        return $1;
    }
    return 0;
}

sub main {
    my $direction = lc $ARGV[0];

    my $workspace_data_json = `i3-msg -t get_workspaces`;
    my $workspace_data = decode_json($workspace_data_json);

    my ($current_ws, $active_output);
    for my $workspace (@$workspace_data) {
        if ($workspace->{focused}) {
            $current_ws = $workspace->{name};
            $active_output = $workspace->{output};
            last;
        }
    }

    my @workspaces_on_output = grep { $_->{output} eq $active_output } @$workspace_data;
    my @sorted_workspaces = sort { extract_number($a->{name}) <=> extract_number($b->{name}) } @workspaces_on_output;

    my $current_num = extract_number($current_ws);
    my $target_ws;

    if ($direction eq 'next') {
        for my $ws (@sorted_workspaces) {
            if (extract_number($ws->{name}) > $current_num) {
                $target_ws = $ws->{name};
                last;
            }
        }

        if (!$target_ws) {
            $target_ws = $sorted_workspaces[0]->{name};
        }
    }
    else {
        my @prev_workspaces = grep { extract_number($_->{name}) < $current_num } @sorted_workspaces;
        if (@prev_workspaces) {
            $target_ws = $prev_workspaces[-1]->{name};
        }
        else {
            $target_ws = $sorted_workspaces[-1]->{name};
        }
    }

    system("i3-msg", "workspace", $target_ws);
}

main() if @ARGV;
